### PR TITLE
arm: mps_an521: Add support for the AN521 on MPS2+

### PR DESCRIPTION
**Pull request by karl-zh**
The AN521 is a Soft Macro Model implementation of the SSE-200 subsystem
with SIE-200 and CMSDK components targeting the MPS2+ board. The
SSE-200 subsystem implements two Cortex-M33 cores.

This PR is based on https://github.com/zephyrproject-rtos/zephyr/pull/6597

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

### CODE REVIEW

File: soc/arm/arm/mps2/arm_mpu_regions.c [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017 Linaro Limited.
+ * Copyright (c) 2018-2019 Linaro Limited.
```
Comment (karl-zh) [inline soc/arm/arm/mps2/arm_mpu_regions.c]: 2017-2019

File: boards/arm/mps2_an521/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+
+ARM MPS2+ AN521 provides the following hardware components:
+
+
+
+- ARM Cortex-M33
+- Soft Macro Model (SMM) implementation of SSE-200 subsystem
+- Memory
+
+  - 16MB internal memory SRAM.
```
Comment (dbkinder) [inline boards/arm/mps2_an521/doc/index.rst]: LGTM.   My only issue would be that list items (such as these) that aren't sentences, shouldn't end with a period

**Review by dbkinder — CHANGES_REQUESTED** [review]
LGTM. My only issue would be that list items that aren't sentences, shouldn't end with a period

File: boards/arm/mps2_an521/doc/index.rst (hunk 2 above)
Comment (karl-zh) [reply boards/arm/mps2_an521/doc/index.rst]: Updated, remove all the "." for the list items.
Thanks.

File: boards/arm/mps2_an521/mps2_an521.dts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	};
+
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "arm,cortex-m33";
+			reg = <0>;
```
Comment (ioannisg) [inline boards/arm/mps2_an521/mps2_an521.dts]: If #12173 is accepted, we might want to add here the mpu configuration (if mpu exists, of course)

File: soc/arm/arm/mps2/arm_mpu_regions.c [hunk 4]
```diff
@@ -9,15 +9,24 @@
 #include <soc.h>
 #include <arch/arm/cortex_m/mpu/arm_mpu.h>
 
+#if defined(CONFIG_CPU_CORTEX_M23) || \
```
Comment (ioannisg) [inline soc/arm/arm/mps2/arm_mpu_regions.c]: ...or you can use the ARCH Kconfigs (ARMV8_M_BASELINE || ARMV8_M_MAINLINE)

File: boards/arm/mps2_an521/mps2_an521_nonsecure_defconfig [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+CONFIG_ARM=y
+CONFIG_SOC_FAMILY_ARM=y
+CONFIG_SOC_SERIES_MPS2=y
+CONFIG_SOC_MPS2_AN521=y
+CONFIG_BOARD_MPS2_AN521=y
+CONFIG_ARM_TRUSTZONE_M=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_RUNTIME_NMI=y
+CONFIG_TRUSTED_EXECUTION_NONSECURE=y
+CONFIG_ARM_MPU=y
```
Comment (ioannisg) [inline boards/arm/mps2_an521/mps2_an521_nonsecure_defconfig]: Why is the MPU not enabled also in the secure board? (Maybe it's somewhere there, but I couldn't find it)

**Review by ioannisg — CHANGES_REQUESTED** [review]
Looks ok, a few comments about MPU

**Review by dbkinder — APPROVED** [review]
docs LGTM, thanks!

**Review by ioannisg — COMMENTED** [review]
Looks good, now, @galak @karl-zh 
I have a suggestion, since #12173 has been merged; you can add MPU information in DTS of the SoC (Board), since this is a Cortex-M33 with non-fixed number of MPU regions.

**Review by ioannisg — APPROVED** [review]

**Comment by ioannisg** [comment]
@karl-zh @galak  should we try to get this in?

**Comment by karl-zh** [comment]
As this patch still not merged and there will be more work about integrate TFM with Zephyr on AN521 after this patch. Mostly are for PSA isolation level one which seems good to have soon. I would close this PR and trace all AN521 supporting stuff on https://github.com/zephyrproject-rtos/zephyr/pull/16450

**Comment by karl-zh** [comment]
Superceded by  #16450

**Comment by galak** [comment]
Let's keep this open and get the base board support merged.

File: soc/arm/arm/mps2/dts_fixup.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2019 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* SoC level DTS fixup file */
+
```
Comment (galak) [inline soc/arm/arm/mps2/dts_fixup.h]: This needs an `#if defined(CONFIG_SOC_MPS2_AN521)` so we don't break MPS2 AN385.

**Review by galak — CHANGES_REQUESTED** [review]
Need to fixup dts_fixup.h so we dont break MPS AN385.

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
> Need to fixup dts_fixup.h so we dont break MPS AN385.

Fixed.

**Review by microbuilder — APPROVED** [review]
lgtm

**Pull request merged**